components/locale_canonicalizer/src/locale_canonicalizer.rs

```diff
@@ -166,228 +205,226 @@ impl LocaleCanonicalizer<'_> {
     /// ```
     ///
     pub fn canonicalize(&self, locale: &mut Locale) -> CanonicalizationResult {
```

Comment:
Question: What does the typical call site look like here?  Is it common that you want to call both maximize for likely subtags as well as canonicalize?

Suggestion 1: If so, perhaps we should make a function that does both.

Suggestion 2: If not (if maximize and canonicalize serve different use cases), then perhaps we should make two entirely separate classes: one that loads likely subtags data and has a maximize function, and the other which loads the canonicalization data and has a canonicalize function.  This has the benefit of making a less monolithic ICU4X API and making it a bit easier to do code and data slicing.  (You could keep the two classes in the same crate.)

Reply:
For the JavaScript usecase, it takes a few jumps, but reading through these:
* https://tc39.es/ecma402/#locale-objects
* https://tc39.es/ecma402/#sec-canonicalizeunicodelocaleid
* https://unicode.org/reports/tr35/#Canonical_Unicode_Locale_Identifiers
* https://unicode.org/reports/tr35/#LocaleId_Canonicalization

shows that canonicalization is applied every time we create an `Intl.Locale` object in JavaScript.

The likely subtags functions `maximize` and `minimize` are members of an `Intl.Locale` instance, so in a sense, yes they are separate. However, `canonicalize` depends upon being able to call `maximize` to handle complex region aliases, so every use of `canonicalize` potentially requires having the likely subtags data available.

We could break this runtime dependency by preprocessing complex regions in the CLDR provider, but that still requires likely subtags data to be available, so we would need some way of specifying a dependency between the alias data transform and the likely subtags data transform. I don't think that is possible right now.

We do use likely subtags operations on their own in Gecko outside of SpiderMonkey, but since we need canonicalization anyway, my thought was to have a singleton `LocaleCanonicalizer` that loads the data once and serves both usecases.

My preference would be to have a single class with three methods, rather than two classes, one of which will need to access the other anyway. If we could break the runtime dependency, then maybe it would make sense to separate them, but even in that case, I think using options to control which data is loaded would probably be fine.

That said, I might be a bit biased by what will make life easier in Gecko, and maybe there are lot of other usecases I'm not considering.

Reply:
> That said, I might be a bit biased by what will make life easier in Gecko, and maybe there are lot of other usecases I'm not considering.

I don't think it's a bad thing. We have a production environment with a business use case and can link it back to a technical decision ahead of us.
In the absence of more business scenarios, having one is dramatically better than having zero :)

I'd be comfortable following your suggestion and if between 0.3 and 1.0 we encounter alternative then refactor.